lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -401,12 +400,40 @@ collectOpsToPipeline(scf::ForOp forOp,
         loadInfo.sharedEncoding =
             getSharedEncIfAllUsersAreDotEnc(loadOp.getResult())
                 .value_or(nullptr);
+
+        // HACK: Triton LLVM codegen has a bug where local_loads from #shared to
+        // #mma layout can lead to invalid code if the loaded shape is smaller
+        // than the mma tile (e.g. loading a 128x1 tensor for an MMAv2 dot with
+        // tile {16,8} is bad because 1 < 8).  To work around this, don't
+        // pipeline such loads.
+        //
+        // The codegen bug is caught by an assertion, so if you think you've
+        // fixed it, feel free to delete this code and see if the assert still
+        // fails.  :)
+        if (!loadInfo.sharedEncoding) {
+          if (auto dotEnc = dot.getResult()
```

Comment:
the condition seems a bit ad hoc to this case. Should we prevent folding of local_load with convert to those mma layout instead?

Reply:
> the condition seems a bit ad hoc to this case

I mean *it* is a bug...

> Should we prevent folding of local_load with convert to those mma layout instead?

That is in fact the first thing I tried.  You need to fix not just the fold but also the creation of "illegal" local_load ops.  It spreads out everywhere.

Reply:
Hi, I'm hitting a codegen issue that seems to be related to this change. In short, an improper shared layout encoding is computed for a load that has multiple uses in software pipelining. E.g,
```

b = tl.load
c = dot(a, b)
d = tl.trans(b)
e = dot(c, d)
```

The layout of the smem buffer for the first dot isn’t exactly the same for the second dot and thus improper swizzling code is generated for one of the dot. I'm not sure about a good fix and would like to hear your thoughts. Thanks!